File: problems/16-bit-manipulation/002-counting-bits/java/Solution.java
public class Solution {
    public int[] countBits(int n) {
        int[] dp = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            // Removing the lowest set bit gives a smaller number
            // whose bit count is already known.
            dp[i] = dp[i & (i - 1)] + 1;
        }

        return dp;
    }
}

File: problems/16-bit-manipulation/002-counting-bits/python/solution.py
class Solution:
    def countBits(self, n: int) -> list[int]:
        dp = [0] * (n + 1)

        for i in range(1, n + 1):
            # Remove the lowest set bit and reuse its answer.
            dp[i] = dp[i & (i - 1)] + 1

        return dp
